src/main/kotlin/org/rust/lang/core/completion/RsKeywordCompletionContributor.kt

```diff
@@ -96,24 +95,48 @@ class RsKeywordCompletionContributor : CompletionContributor(), DumbAware {
         RsPsiPattern.inAnyLoop.and(newCodeStatementPattern())
 
     private fun baseDeclarationPattern(): PsiElementPattern.Capture<PsiElement> =
-        psiElement<PsiElement>().andOr(
-            psiElement().withParent(RsPath::class.java),
-            psiElement().withParent(or(psiElement<RsModItem>(), psiElement<RsFile>()))
-        )
+        psiElement()
+            .withParent(or(psiElement<RsPath>(), psiElement<RsModItem>(), psiElement<RsFile>()))
 
     private fun baseCodeStatementPattern(): PsiElementPattern.Capture<PsiElement> =
-        psiElement<PsiElement>()
+        psiElement()
             .inside(psiElement<RsFunction>())
             .andNot(psiElement().withParent(RsModItem::class.java))
 
     private fun statementBeginningPattern(vararg startWords: String): PsiElementPattern.Capture<PsiElement> =
-        psiElement<PsiElement>()
-            .withElementType(TokenSet.create(IDENTIFIER))
-            .and(RsPsiPattern.onStatementBeginning(*startWords))
+        psiElement(IDENTIFIER).and(RsPsiPattern.onStatementBeginning(*startWords))
 
     private fun elsePattern(): PsiElementPattern.Capture<PsiElement> {
         val braceAfterIf = psiElement(RBRACE).withSuperParent(2, psiElement(IF_EXPR))
-        return psiElement<PsiElement>()
-            .afterLeafSkipping(psiElement(PsiWhiteSpace::class.java), braceAfterIf)
+        return psiElement().afterLeafSkipping(RsPsiPattern.whitespace, braceAfterIf)
+    }
+
+    private fun wherePattern(): PsiElementPattern.Capture<PsiElement> {
+        val typeParameters = psiElement<RsTypeParameterList>()
+
+        val function = psiElement<RsFunction>()
+            .withLastChildSkipping(RsPsiPattern.error, or(psiElement<RsValueParameterList>(), psiElement<RsRetType>()))
+            .andOr(psiElement().withChild(psiElement<RsTypeParameterList>()),
+                psiElement().withParent(RsTraitOrImpl::class.java))
+
+        val struct = psiElement<RsStructItem>()
+            .withChild(typeParameters)
+            .withLastChildSkipping(RsPsiPattern.error, or(typeParameters, psiElement<RsTupleFields>()))
+
+        val enum = psiElement<RsEnumItem>()
+            .withLastChildSkipping(RsPsiPattern.error, typeParameters)
+
+        val typeAlias = psiElement<RsTypeAlias>()
+            .withLastChildSkipping(RsPsiPattern.error, typeParameters)
+            .andNot(psiElement().withParent(RsTraitOrImpl::class.java))
+
+        val trait = psiElement<RsTraitItem>()
+            .withLastChildSkipping(RsPsiPattern.error, or(psiElement(IDENTIFIER), typeParameters))
+
+        val impl = psiElement<RsImplItem>()
+            .withLastChildSkipping(RsPsiPattern.error, psiElement<RsTypeReference>())
+
+        return psiElement()
+            .withPrevSiblingSkipping(RsPsiPattern.whitespace, or(function, struct, enum, typeAlias, trait, impl))
```

Comment:
Ideally, we should use parser to provide keyword completion, like Kotlin does: https://github.com/JetBrains/kotlin/blob/master/idea/idea-completion/src/org/jetbrains/kotlin/idea/completion/KeywordCompletion.kt#L199.

But pattern based approach should work quite well in practice!